Added CustomExport functionality and updated Demo

hbermani opened the pull request:
## Related Issue

Relate the Github issue 
https://github.com/mbrn/material-table/pull/2089

## Description

Added functionality to create customExport function. This is very helpful when exporting nested/complex objects . I also updated the demo to show customExport component in action.


## Impacted Areas in Application

List general components of the application that this PR will affect:
- Exporting data


## Additional Notes

This is my first ever contribution to an open source project, please advise if I need to do anything differently.

hbermani committed:
Added CustomExport funcationality and updated Demo

matthewoestreich commented:
Hey @hbermani thank you! Unfortunately, [this has already been implemented](https://material-table-core.com/docs/api/props/options#exportmenu) =/

hbermani commented:
> Hey @hbermani thank you! Unfortunately, [this has already been implemented](https://material-table-core.com/docs/api/props/options#exportmenu) =/

May be it's just me as my brain is fried after a full day, but what this allows you to do is specify the customExport function within exporting to .csv / pdf as it is changing the datas content. This is very handy when exporting complex objects so they don't show as [object Object].

matthewoestreich commented:
Lol no worries. I've been known to misunderstand. 

If I do understand this correctly, you are wanting to specify a custom function for exporting data, correct?

In the new model, the concept of exporting csv or exporting pdf no longer exists. You provide an array of objects with this shape:

```js
exportMenu: [{
      label: 'Export PDF',
      exportFunc: (cols, datas) => ExportPdf(cols, datas, 'myPdfFileName')
    },
//....
```
And we add it to the menu. Notice the exportFunc - I believe that accomplishes the same thing as you described. 

This way allows is to not hard code csv or pdf. Just give us an array of objects with the correct shape and we add it to the menu.

We separated the pdf and csv export defaults due to bundle bloat. That article shows how you can still use them though. You just have to import them.

matthewoestreich commented:
@hbermani 

To demonstrate, this is how you could use a custom export function to detect if the column header is not a string (if you're using JSX as the column title). This way you can replace it with a string so that the export doesn't show `[object, Object]` as the column header.

Or you could use the default `import ExportCsv from '@material-table/core/exporters/csv';`

```js
const App = () => {
  return (
    <MaterialTable
      // ... other props
      options={{
        exportMenu: [
          {
            label: "Export CSV",
            exportFunc: (columns, data) => {
              // Turn headers into array of strings
              const headerRow = columns.map((col) => {
                if (typeof col.title === "object") {
                  // I am not sure what props the Translate component exposes
                  // but you would need to change `text` in `col.title.props.text`
                  // to whatever prop you need.
                  return col.title.props.text;
                }
                return col.title;
              });

              // Turn data into an array of string arrays, without the `tableData` prop
              const dataRows = data.map(({ tableData, ...row }) => Object.values(row));

              // Aggregate header data and 'body' data
              // Mirror default export behavior by joining data via
              // the delimiter specified within material table (by default comma delimited)
              const { exportDelimiter } = tableRef.current.props.options;
              const delimiter = exportDelimiter ? exportDelimiter : ",";
              const csvContent = [headerRow, ...dataRows].map((e) => e.join(delimiter)).join("\n");

              // This mirrors the default export behavior where the
              // exported file name is the table title.
              const csvFileName = tableRef.current.props.title;

              // Allow user to download file as .csv
              downloadCsv(csvContent, csvFileName);
            },
          },
        ],
      }}
    />
  );
};
```

hbermani commented:
> Lol no worries. I've been known to misunderstand.
> 
> If I do understand this correctly, you are wanting to specify a custom function for exporting data, correct?
> 
> In the new model, the concept of exporting csv or exporting pdf no longer exists. You provide an array of objects with this shape:
> 
> ```js
> exportMenu: [{
>       label: 'Export PDF',
>       exportFunc: (cols, datas) => ExportPdf(cols, datas, 'myPdfFileName')
>     },
> //....
> ```
> 
> And we add it to the menu. Notice the exportFunc - I believe that accomplishes the same thing as you described.
> 
> This way allows is to not hard code csv or pdf. Just give us an array of objects with the correct shape and we add it to the menu.
> 
> We separated the pdf and csv export defaults due to bundle bloat. That article shows how you can still use them though. You just have to import them.

I absolutely love how you guys separated the export menu like that, makes sense. Having said that the challenge here is that datas (referenced above) doesn't always hold the correct shape (i.e. all columns are not always flat data), that's why the customExport function at Column Definition level handles that then passes it to the exportToCSV / pdf etc.. This is particularly useful in scenarios where you might have complex user data or embedded arrays within a column, as show in updated demo file.

hbermani commented:
> @hbermani
> 
> To demonstrate, this is how you could use a custom export function to detect if the column header is not a string (if you're using JSX as the column title). This way you can replace it with a string so that the export doesn't show `[object, Object]` as the column header.
> 
> Or you could use the default `import ExportCsv from '@material-table/core/exporters/csv';`
> 
> ```js
> const App = () => {
>   return (
>     <MaterialTable
>       // ... other props
>       options={{
>         exportMenu: [
>           {
>             label: "Export CSV",
>             exportFunc: (columns, data) => {
>               // Turn headers into array of strings
>               const headerRow = columns.map((col) => {
>                 if (typeof col.title === "object") {
>                   // I am not sure what props the Translate component exposes
>                   // but you would need to change `text` in `col.title.props.text`
>                   // to whatever prop you need.
>                   return col.title.props.text;
>                 }
>                 return col.title;
>               });
> 
>               // Turn data into an array of string arrays, without the `tableData` prop
>               const dataRows = data.map(({ tableData, ...row }) => Object.values(row));
> 
>               // Aggregate header data and 'body' data
>               // Mirror default export behavior by joining data via
>               // the delimiter specified within material table (by default comma delimited)
>               const { exportDelimiter } = tableRef.current.props.options;
>               const delimiter = exportDelimiter ? exportDelimiter : ",";
>               const csvContent = [headerRow, ...dataRows].map((e) => e.join(delimiter)).join("\n");
> 
>               // This mirrors the default export behavior where the
>               // exported file name is the table title.
>               const csvFileName = tableRef.current.props.title;
> 
>               // Allow user to download file as .csv
>               downloadCsv(csvContent, csvFileName);
>             },
>           },
>         ],
>       }}
>     />
>   );
> };
> ```

That should do it, but imagine with the complexity of so many different data structures in an application how many exporter conditions you would have. I don't think it would scale nicely. The export to csv should do one thing only (export to csv), custom manipulation of underlying data is better managed IMHO at the column level.

matthewoestreich commented:
@hbermani but thats the thing though. @material-table/core has no concept of "csv". It has no default way for you to export a CSV. 

In fact, it doesn't matter what you're exporting (which allows this to scale 100x better than before). Let's say you need to export table data to a custom file format (for whatever reason), as well as to pdf...

We have abstracted the export process - exportCsv no longer exists. We give you a function and the data to do what you need to do, PLUS the freedom to add custom export menu items (before you had two options only exportCsv and exportPdf, which were hard coded and bloating the bundle, now you are not limited to any of that).

Hope this makes sense.

hbermani commented:
> @hbermani but thats the thing though. @material-table/core has no concept of "csv". It has no default way for you to export a CSV.
> 
> In fact, it doesn't matter what you're exporting. Let's say you need to export table data to a custom file format (for whatever reason), as well as to pdf...
> 
> We have abstracted the export process - exportCsv no longer exists. We give you a function and the data to do what you need to do, PLUS the freedom to add custom export menu items (before you had two options only exportCsv and exportPdf, now you are not limited to any of that).
> 
> Hope this makes sense.

It makes total sense, so picture the data structure below, if we wanted to export the field 'teams' into csv, pdf, txt etc alongside other columns, the current implementation means, we need to reimplement exportFunc for each exporter rather than the proposed addition which allows further flexibility as it handles data manipulation at Column Definition thus avoiding duplication, it's also more efficient as it doesn't have to map through the data again as it's handles the changes when it's getting the field value. 
 
```
const global_data_CustomExport = [
  {
    name: 'Mehmet',
    surname: 'Baran',
    birthYear: 1987,
    birthCity: 63,
    teams: ['Team A', 'Team B']
  },
  {
    name: 'Zerya Betül',
    surname: 'Baran',
    birthYear: 2017,
    birthCity: 34,
    teams: ['Team C', 'Team D', 'Team E']
  }
]```

matthewoestreich commented:
I still don't see the difference. You are not 'reimplementing' the func for each menu item. Each menu item *has* to have it's own func. 

Inside of said func, you can literally do whatever you need to do with data from anywhere you want to get it from. Literally. You can do whatever you want. You have complete control of the data that is being exported.

matthewoestreich commented:
You could just do:

I believe including that option as part of the column definition only confuses writing code for this library, as well as writing code using this library. It doesn't make sense to have it as part of the column definition.

```js
exportMenu: [{
      label: 'Export PDF',
      exportFunc: (cols, datas) => { 
        // You can do whatever you want here
        let dataToExport = [];
        cols.map(col => { col.teams.forEach(team => dataToExport.push(team) ) }) }
    },
//....
```

hbermani commented:
> I still don't see the difference. You are not 'reimplementing' the func for each menu item. Each menu item _has_ to have it's own func.
> 
> Inside of said func, you can literally do whatever you need to do with data from anywhere you want to get it from. Literally. You can do whatever you want. You have complete control of the data that is being exported.

That's the thing, I am not proposing a change in how the export logic has been implemented, it's great as is. The change is mainly concerned with the underlying data before it gets to whatever custom Exporter. To be more specific this is the bit of code which handles it under  src/components/MTableToolbar/index.js

```
 const data = (props.exportAllData ? props.data : props.renderData).map(
      (rowData) =>
        columns.map((columnDef) => {
          if (typeof columnDef.customExport === 'function') {
            return columnDef.customExport(rowData);
          }
          return props.getFieldValue(rowData, columnDef);
        })
```
The type check for a "function" above allows users to pass a custom function at column definition level that handles the shaping of the data before it's passed to whatever exporter being used rather than looping data at each exporter to shape it.

hbermani commented:
> You could just do:
> 
> I believe including that option as part of the column definition only confuses writing code for this library, as well as writing code using this library. It doesn't make sense to have it as part of the column definition.
> 
> ```js
> exportMenu: [{
>       label: 'Export PDF',
>       exportFunc: (cols, datas) => { 
>         // You can do whatever you want here
>         let dataToExport = [];
>         cols.map(col => { col.teams.forEach(team => dataToExport.push(team) ) }) }
>     },
> //....
> ```

Now let's say you want to export this to a flat file or csv too.... you will have to re-write the same logic for each exporter. 
``` 
         let dataToExport = [];
>         cols.map(col => { col.teams.forEach(team => dataToExport.push(team) ) }) }
```

matthewoestreich commented:
That's exactly what exportFunc does. I just showed a demo that checks for the header to be an object (instead of a func) but its the exact same logic. 

Then we have a property on a columns object for exporting data as well as properties to export data in options. Having to configure two separate settings to accomplish the same thing makes no sense to me.

matthewoestreich commented:
Exactly that's what we want. It allows for freedom of exporting to whatever you want. Even a flat file with any extension you could possibly imagine. 

That's exactly why we provide the default csv and pdf exports. To keep your code clean. 

Outside of that, anything custom would need to have it's own method used to export. 

Adding another prop only makes things a million times more confusing. It's just unnecessary. There are ways you can abstract your exporters to eliminate duplicate code, but all of that needs to be done in userland in the export func. This allows us to remain highly configurable, easy to understand, and efficient.

hbermani commented:
> That's exactly what exportFunc does. I just showed a demo that checks for the header to be an object (instead of a func) but its the exact same logic.
> 
> Then we have a property on a columns object for exporting data as well as properties to export data in options. Having to configure two separate settings to accomplish the same thing makes no sense to me.

Our brains are clearly wired differently :-D . I can see it's the same logic, it's just that extra bit of flexibility so the logic is handled at the data layer (column) rather than the exporter, otherwise with many export functions you will end up repeating the same logic and in turn have less consistency between exporters (Which in all fairness may be desirable by some), but in other scenarios isn't. 

I feel we are going in circles here lol!

matthewoestreich commented:
I've been refactoring this code and one of the biggest problems is unnecessary props like that. Which, are not easy to troubleshoot. I can only imagine the pain of trying to troubleshoot the fact there are 2 properties in diff places that help serve a single feature. Its one of those things that if it isn't documented or commented - have fun reading that code. 

One of the major benefits of material-table is how we expose so much of our internals, allowing you to be extremely modular. 

I like to keep it simple, above all. Even if that means "duplicate code". I like to stay away from "magic" like that. It's not obvious and is hard to understand.

hbermani commented:
The reason I proposed it at the column level is because custom render is handled there as well and the two are very interlinked in terms of what they serve, so to render a complex object using the "render" prop under Columns to me is no different to using "customExport" or it can be named whatever to handle the same object in the exporter. One is prepping it for HTML the other for whatever format the exporter likes. 

And the best part is, it doesn't change any of the existing implementation, it just adds additional flexibility, as an option, for scenarios where it makes sense and consistency is desired across exporters.

matthewoestreich commented:
I would agree that we are wired differently lol.. To me, custom render and custom export are not, and should not be, reliant on each other whatsoever. Coupling them is only asking for trouble.

The custom render function for columns makes sense to have on each column object, in my opinion. That is because it has to deal directly with the column itself. Exporting does not deal with a specific column (although, it *could* now) - it deals with exporting data.

Ultimately, using exportMenu gives you access to the exact same columns object/data anyway. If one of our concerns is duplicate code, why handle the same feature (exporting) in two different (duplicate) places? That should be concerning to us. (again, especially when the exportFunc provides you access to that exact same data anyway. handling this in the columns object feels convoluted).

hbermani commented:
> I would agree that we are wired differently lol.. To me, custom render and custom export are not, and should not be, reliant on each other whatsoever. Coupling them is only asking for trouble.
> 
> The custom render function for columns makes sense to have on each column object, in my opinion. That is because it has to deal directly with the column itself. Exporting does not deal with a specific column (although, it _could_ now) - it deals with exporting data.
> 
> Ultimately, using exportMenu gives you access to the exact same columns object/data anyway. If one of our concerns is duplicate code, why handle the same feature (exporting) in two different (duplicate) places? That should be concerning to us. (again, especially when the exportFunc provides you access to that exact same data anyway. handling this in the columns object feels convoluted).

I certainly don't want them to be coupled... blimey that would be a nightmare indeed lol... I am saying they are similar in terms of where they are placed in the code. And to the same point you raised about handling specific column, this is exactly what we want to do here, handle a specific column that does not hold flat data and to me the data logic shouldn't be handled in the exporter.  


So using my not so great skills of trying to illustrate it (haha) here is how I picture it.

Rendering:
-  data  ==> Column render prop (Changes content to suit HTML) => Displays on table

Exporting (current implementation) :
- data ==> Displays on Exporter (Handles data change there)

Export (Proposed add-on)
In addition to the above
- data ==> Column export prop (Changes content to suit export) => Displays on Exporter 

May be the use of the term customExport is not the best, what we are trying to do here is essentially ETL for export purposes.

matthewoestreich commented:
I completely understand what you're saying but to me it would be so much easier to read with everything in one spot. The exporting is what takes the raw data, manipulates it, then exports it. 

Otherwise it's like you would have to scroll to your columns (or go to the columns file), edit the export method then go back into the export func to use it. It's not clear to someone reading that code that the data is manipulated beforehand. Intent is not very clear. 

Your way would 100% work, I just don't like it lol (much like you don't like mine haha)... 

With that being said I'm open to changing it, if that is what everyone agrees upon. I think taking a poll of some sort would help. Perhaps, I am the odd man out?

hbermani commented:
> I completely understand what you're saying but to me it would be so much easier to read with everything in one spot. The exporting is what takes the raw data, manipulates it, then exports it.
> 
> Your way would 100% work, I just don't like it lol (much like you don't like mine haha)...
> 
> With that being said I'm open to changing it, if that is what everyone agrees upon. I think taking a poll of some sort would help. Perhaps, I am the odd man out?

Haha, I honestly don't see it as either or, as it an addition rather than a replacement to existing implementation. I'll post on the original PR for people to add their thoughts and link to here. :-)

matthewoestreich commented:
Another benefit that the freedom of exportMenu allows is exporting remote data. Otherwise, there would have to be some hard coded prop for exporting remote data. 

https://github.com/material-table-core/core/issues/25#issuecomment-767046324

Overall, to me, it's just easier to understand as a consumer as well as a contributor. It's just more obvious and the abstraction helps other aspects and features of the app.

matthewoestreich commented:
I think we could change the prop name exportFunc to onExport though lol. I hate myself for not thinking of that sooner.

hbermani commented:
> I think we could change the prop name exportFunc to onExport though lol. I hate myself for not thinking of that sooner.

Oooo... though that would cause breaking changes...how about if the new customExport is called onExport ... I know I am blowing your mind now haha!!

matthewoestreich commented:
lol @Domino987 what do you think about this? I know it's a long read but whatever you decide is where we will go. You are the deciding vote, my friend!

My vote is to not introduce this property as I believe it confuses things, both internally and externally, and it is unnecessary. 

Thoughts?

hbermani commented:
> lol @Domino987 what do you think about this? I know it's a long read but whatever you decide is where we will go. You are the deciding vote, my friend!
> 
> My vote is to not introduce this property as I believe it confuses things, both internally and externally, and it is unnecessary.
> 
> Thoughts?

@Domino987 Summary of my thoughts here https://github.com/mbrn/material-table/pull/2089#issuecomment-767100416

Domino987 commented:
Ok, people ;) That's a long thread and I see both sides. You have both great arguments and I have a difficulty to decide, but I think I would go with @hbermani 's way I think.

Material table is not only offering great customization, which is totally does, but I think the main reason is it's easy use.

In my estimation, 99% of the use cases will be I want to export to csv/pdf and I only have this one header that needs to change (at least that was the one use case I had a while back).

That should not lead to writing one or more whole data manipulation and export functions, IMO.

IF somebody wants a custom export in addition, they could just use a free action and handle the onClick themselves right?. At least that's how I fixed it. And this will probably not be the use case,  I think.

SO while I see both sides, I think the majority of use cases will be single column header changes, or flattening of one column, so my vote would go to single column.

Domino987 commented on types/index.d.ts:
```diff
@@ -162,6 +162,7 @@ export interface Column<RowData extends object> {
     data2: RowData,
     type: 'row' | 'group'
   ) => number;
+  customExport?: (rowData: RowData) => void;
```
This return value should not be void, but instead unknown, so anything could be returned.

matthewoestreich commented:
Right on, if you all think it is beneficial then lets add it.  I still think it is unnecessary.

Whether one column or 100 columns - simplicity matters over all, and this way is not very simple (yes, it is simple for a very specific use case, which again, we should have no opinion on). Imagine having to make that change on 100 different individual columns....it makes no sense to me lol

Another example of how this can be harmful - sometimes it is hard to tell at face value what the purpose of a property is.. I feel like this solution is how we wind up with things like this:

https://github.com/material-table-core/core/blob/881a48a9d19d10b28fb3522a7f39e9a70fb7e5b7/src/utils/data-manager.js#L960

and

https://github.com/material-table-core/core/blob/881a48a9d19d10b28fb3522a7f39e9a70fb7e5b7/src/utils/data-manager.js#L539

What is the difference (at face value)? 

Which one should I use for what and why? 

Now I have to not only read code, but try and understand the context that code is being used in, all to simply understand the difference between those two. (Yes, this is a simple example, but the idea remains the same.)

Separating functionality like this is the definition of spaghetti code, and it's how we wound up with so many confusing props strewn about everywhere to begin with... This code base is extremely difficult to read because of things like this, imo.. 
You're right, adding a prop is quick and easy for a single column, but it's a bandaid solution for a specific use case, and no easier than adding a single object to exportMenu.

Just my 2c tho, feel free to make any changes!

matthewoestreich commented:
Let me know if you publish a new version with these changes - I'll update the docs.

hbermani commented:
Good morning folks, from the UK, please let me know if I need to do anything my end to get this past the finish line. 

@oze4 , thank you for your patience in considering this PR, replacing the term customExport with onExport might be more intuitive, whichever you feel makes more sense to add further clarity.

mogilvie commented:
My use case was exactly as @Domino987 above. One or two columns, not wholesale manipulation. The heavy lifting of full exports was handled at the back end as the formatting and manipulation of the exported document required additional processing (more data than just the display data / logo's formatting / watermarks etc)
Thanks for the consideration, and cheers @hbermani for the PR.

Domino987 commented:
> Good morning folks, from the UK, please let me know if I need to do anything my end to get this past the finish line.
> 
> @oze4 , thank you for your patience in considering this PR, replacing the term customExport with onExport might be more intuitive, whichever you feel makes more sense to add further clarity.

I added one change request.

matthewoestreich commented:
>One or two columns, not wholesale manipulation

@mogilvie it isn't "wholesale manipulation" and is very very simple to make necessary changes (and not as confusing imo). Keeps everything in one place!

Have you read the codebase by any chance @mogilvie ? Props spread out like this make it incredibly difficult to digest this codebase.

Just needing to edit one or two columns is still a specific use case. Our library should not be coded for specific use cases it literally goes against the grain.

matthewoestreich commented:
If you search through mbrn/material-table issues you will see people wanting more than csv or pdf exports. exportMenu allows you to export however you want and now you're not limited. Exporting should not be assumed or handled by material-table (or the "back end").  Boggles my mind you all cant see that!!

hbermani committed:
Updated customExport prop to return unknown

hbermani commented:
> If you search through mbrn/material-table issues you will see people wanting more than csv or pdf exports. exportMenu allows you to export however you want and now you're not limited. Exporting should not be assumed or handled by material-table (or the "back end"). Boggles my mind you all cant see that!!

I can certainly see your point Oze4, but it comes down to ease of use to manipulate the data in one place, just like we handle custom render for complex columns.  Besides, it doesn't stop someone from applying changes at an exporter level, so it's the best of both worlds.

hbermani commented:
> > Good morning folks, from the UK, please let me know if I need to do anything my end to get this past the finish line.
> > @oze4 , thank you for your patience in considering this PR, replacing the term customExport with onExport might be more intuitive, whichever you feel makes more sense to add further clarity.
> 
> I added one change request.

I have updated prop return type.

matthewoestreich commented:
@hbermani I completely understand the intent. It just doesn't make much sense to me. Render !== export so they should be treated differently, but as you say this doesn't prevent someone from doing anything it just gives them more options. 

But again, if you read through the codebase you'll quickly discover props spread out across components like this (that serve a single feature) make things extremely difficult to keep track of and digest. So now when someone is using this codebase it isn't as straight forward. They will see an `onExport` in columns and an `onExport` in `exportMenu`. See what I'm saying? Two props. Named the exact same. Serving the exact same purpose. Yet they are in two completely different places, it's just confusing and unnecessary. 

Above everything else, that is my biggest concern. Not use case, not ease-of-use (both of these solutions are easy), but the fact that it makes the codebase feel like spaghetti, hard to read, and creates duplicate props..

hbermani commented:
@oze4 Isn't that what docs are for though, personally I just refer to docs like the ones here (https://material-table.com/#/docs/all-props) and read what it does under each section as things change all the time. I kept the naming as customExport for now. 

There isn't more I can say about this, if you are the gatekeeper and you don't want it in then it's your decision and nothing more I can say or do will change that. I appreciate the time you took to review this.

matthewoestreich commented:
Docs can help in any situation but relying on them isn't a good idea. Your code should be "self documenting" and easy to understand why a certain prop/func/etc.. is being used based off of context and naming. The ONLY way to know the column.onExport and the exportMenu.onExport are related is via the docs. That's a huge problem. And is exactly why performing "magic" like this shouldn't be implemented. 

Not to mention the extra worked that has now been created. Someone has to update the docs. What if that prop changes name? Now you have to edit 2 files, updates the docs, etc.. etc.. etc...

Now lets say we are relying on the docs to explain 20 props. This gets out of hand very quickly (just read the current codebase for yourself to see).

hbermani commented:
Do you want me to add comments on each new line I added to explain what it does for future code reviews? 

Ps. it's still called customExport and not onExport in column.

matthewoestreich commented:
I think it wouldn't hurt but it's up to you. 

The prop names haven't been changed yet (just imagine how much more confusing it is to relate columns.customExport to exportMenu.exportFunc). At least naming them the other way is a little more obvious. That's why I used those names in my examples.

hbermani committed:
Added comments to explain code

hbermani committed:
Added comments for further explanation

hbermani commented:
Added the comments, would you like me to amend anything else or is it still a no go?

matthewoestreich commented:
I think its a for sure go (I am the odd one out here).

I have no other requests. Unless @Domino987 has any suggestions I think we are good to merge this one.

hbermani commented:
> I think its a for sure go (I am the odd one out here).
> 
> I have no other requests. Unless @Domino987 has any suggestions I think we are good to merge this one.

Thank you, appreciate your patience with it.

matthewoestreich commented:
I may have strong opinions but let's be honest I should be thanking you all for your patience lol

Thanks for the contribution!

hbermani commented:
> I may have strong opinions but let's be honest I should be thanking you all for your patience lol
> 
> Thanks for the contribution!

No thank you for keeping this much needed library alive!!

matthewoestreich commented:
I appreciate the gesture but it's not just me. I'm like 1/1000th lol it's really bc of all of us, including you, that are keeping it alive. @Domino987 deserves all the kudos.

matthewoestreich commented:
@Domino987 are we good to merge this?

Domino987 committed:
Merge branch 'master' into master

Domino987 commented:
Yes looks good to me. Thank you for your contribution. It is much appreciated. Sorry for being afk a few days.

matthewoestreich commented:
This has been published with v2.3.25!

Thanks @Domino987 appreciate it!
